doc_internal/fish-riir-plan.md

```diff
@@ -0,0 +1,77 @@
+
+We will not use tokio, serde, async, or other fancy Rust frameworks initially.
+
+### FFI
+
+Rust/C++ interop will use [autocxx](https://github.com/google/autocxx), [Cxx](https://cxx.rs), and possibly [bindgen](https://rust-lang.github.io/rust-bindgen/). I've forked these for fish (see the Development Guide). Once the port is done, we will stop using them, except perhaps bindgen for PCRE2.
+
+We will use [corrosion](https://github.com/corrosion-rs/corrosion) for CMake integration.
+
+Inefficiencies (e.g. extra string copying) at the FFI layer are fine, since it will all get thrown away.
+
+Tests can stay in fish_tests.cpp or be moved into Rust .rs files; either is fine.
+
+### Strings
+
+Rust's `String` / `&str` types cannot represent non-UTF8 filenames or data using the default encoding scheme. That's why all string conversions must go through fish's encoding scheme (using the private-use area to encode invalid sequences). For example, fish cannot use `File::open` with a `&str` because the decoding will be incorrect.
+
+So instead of `String`, fish will use its own string type, and manage encoding and decoding as it does today. However we will make some specific changes:
+
+1. Drop the nul-terminated requirement. When passing `const wchar_t*` back to C++, we will allocate and copy into a nul-terminated buffer.
+2. Drop support for 16-bit wchar. fish will use UTF32 on all platforms, and manage conversions itself.
+
+After the port we can consider moving to UTF-8, for memory usage reasons.
+
+See the Rust Development Guide for more on strings.
+
+### Thread Safety
+
+Allowing [background functions](https://github.com/fish-shell/fish-shell/issues/238) and concurrent functions has been a goal for many years. I have been nursing [a long-lived branch](https://github.com/ridiculousfish/fish-shell/tree/concurrent_even_simpler) which allows full threaded execution. But though the changes are small, I have been reluctant to propose them, because they will make reasoning about the shell internals too complex: it is difficult in C++ to check and enforce what crosses thread boundaries.
+
+This is Rust's bread and butter: we will encode thread requirements into our types, making it explicit and compiler-checked, via Send and Sync. Rust will allow turning on concurrent mode in a safe way, with a manageable increase in complexity, finally enabling this feature.
+
+## Timeline
+
+Handwaving, 6 months? Frankly unknown - there's 102 remaining .cpp files of various lengths. It'll go faster as we get better at it. Peter (ridiculous_fish) is motivated to work on this, other current contributors have some Rust as well, and we may also get new contributors from the Rust community. Part of the point is to make contribution easier.
```

Comment:
I'm eager to contribute as well, probably at modest capacity.
Seems like you have already overcome the biggest initial hurdles.
I'm always motivated to help people get up to speed with fish and Rust.

Reply:
It took me 3 hours to [port util.cpp](https://github.com/ridiculousfish/fish-shell/pull/2), which has 200 lines.
With my new knowledge I could probably do it in 1 hour.
Extrapolating the 1-hour estimate to the remaining 60k lines gives us some 300 hours of work.
I'm sure that other modules are harder and might take longer but this number leaves me optimistic.